packages/middleware-bucket-endpoint/src/bucketEndpointMiddleware.ts

```diff
@@ -0,0 +1,70 @@
+import { bucketHostname } from "./bucketHostname";
+import { BucketEndpointResolvedConfig } from "./configurations";
+import {
+  BuildHandler,
+  BuildHandlerArguments,
+  BuildHandlerOptions,
+  BuildHandlerOutput,
+  BuildMiddleware,
+  MetadataBearer,
+  Pluggable
+} from "@aws-sdk/types";
+import { HttpRequest } from "@aws-sdk/protocol-http";
+
+export function bucketEndpointMiddleware(
+  options: BucketEndpointResolvedConfig
+): BuildMiddleware<any, any> {
+  return <Output extends MetadataBearer>(
+    next: BuildHandler<any, Output>
+  ): BuildHandler<any, Output> => async (
+    args: BuildHandlerArguments<any>
+  ): Promise<BuildHandlerOutput<Output>> => {
+    const { Bucket: bucketName } = args.input;
+    let replaceBucketInPath =
+      options.preformedBucketEndpoint || options.bucketEndpoint;
+    let request = args.request;
+    if (HttpRequest.isInstance(request)) {
+      if (options.bucketEndpoint) {
+        request.hostname = bucketName;
+      } else if (!options.preformedBucketEndpoint) {
+        const { hostname, bucketEndpoint } = bucketHostname({
+          bucketName,
+          baseHostname: request.hostname,
+          accelerateEndpoint: options.useAccelerateEndpoint,
+          dualstackEndpoint: options.useDualstackEndpoint,
+          pathStyleEndpoint: options.forcePathStyle,
+          sslCompatible: request.protocol === "https:"
+        });
+
+        request.hostname = hostname;
+        replaceBucketInPath = bucketEndpoint;
+      }
+
+      if (replaceBucketInPath) {
+        request.path = request.path.replace(/^(\/)?[^\/]+/, "");
+        if (request.path === "") {
+          request.path = "/";
+        }
```

Comment:
What's the rationale for removing the path? I think we need more doc/comments on this.

Reply:
Reading it, it's removing the bucket name from the path as the second-part of transforming from path-style to hostname-style bucket possibly started by `bucketHostname()`, e.g. from `s3.amazonaws.com/my-bucket/my-key` to `my-bucket.s3.amazonaws.com/my-key`, possibly by whatever happened before this middleware.

What's unclear to me at this point is why is the bucket name already in the request path? My understanding was that [the path-style was being deprecated](https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/), so I would expect the request path to already only be the key.

Reply:
@simonbuchan Even though path-style is deprecated, the API model shipped by S3 service team is still path-style. So SDK needs to do the heavy lifting to move the bucket name from path to the hostname.